File: Novels/packages/5-mobile-ios/StoryReader/Repository/ReadingChallengesRepository.swift
import Foundation
import Combine

final class ReadingChallengesRepository {
    private let graphQLService = GraphQLService.shared
    
    func getChallengeProgress(challengeId: String) -> AnyPublisher<ChallengeProgressPayload, Error> {
        graphQLService.getChallengeProgress(challengeId: challengeId)
    }
    
    func getFriendProgress(challengeId: String) -> AnyPublisher<[ChallengeParticipant], Error> {
        graphQLService.getFriendChallengeProgress(challengeId: challengeId)
    }
    
    func updateProgress(challengeId: String, progress: Int) -> AnyPublisher<Bool, Error> {
        graphQLService.updateReadingChallengeProgress(challengeId: challengeId, progress: progress)
    }
    
    func setReadingGoal(goalType: String, target: Int, timeRange: String, startDate: String, endDate: String) -> AnyPublisher<ReadingGoal, Error> {
        graphQLService.setReadingGoal(
            goalType: goalType,
            target: target,
            timeRange: timeRange,
            startDate: startDate,
            endDate: endDate
        )
    }
    
    func getActivityFeed(page: Int = 1, limit: Int = 20) -> AnyPublisher<ActivityFeedPage, Error> {
        graphQLService.getActivityFeed(page: page, limit: limit)
    }
    
    func getReadingStatistics() -> AnyPublisher<ReadingStatistics, Error> {
        graphQLService.getReadingStatistics()
    }
}

File: Novels/packages/6-mobile-android/StoryReader/app/src/main/java/com/storysphere/storyreader/repository/ReadingChallengesRepository.kt
package com.storysphere.storyreader.repository

import com.storysphere.storyreader.model.ActivityFeed
import com.storysphere.storyreader.model.ChallengeProgress
import com.storysphere.storyreader.model.ReadingGoal
import com.storysphere.storyreader.model.ReadingStatistics
import com.storysphere.storyreader.network.GraphQLService
import javax.inject.Inject
import javax.inject.Singleton

@Singleton
class ReadingChallengesRepository @Inject constructor(
    private val graphQLService: GraphQLService
) {
    suspend fun getChallengeProgress(challengeId: String): ChallengeProgress =
        graphQLService.getChallengeProgress(challengeId).getOrThrow()

    suspend fun getFriendProgress(challengeId: String) =
        graphQLService.getFriendChallengeProgress(challengeId).getOrThrow()

    suspend fun updateProgress(challengeId: String, progress: Int): Boolean =
        graphQLService.updateReadingChallengeProgress(challengeId, progress).getOrThrow()

    suspend fun setReadingGoal(
        goalType: String,
        target: Int,
        timeRange: String,
        startDate: String,
        endDate: String
    ): ReadingGoal = graphQLService
        .setReadingGoal(goalType, target, timeRange, startDate, endDate)
        .getOrThrow()

    suspend fun getActivityFeed(page: Int = 1, limit: Int = 20): ActivityFeed =
        graphQLService.getActivityFeed(page, limit).getOrThrow()

    suspend fun getReadingStatistics(): ReadingStatistics =
        graphQLService.getReadingStatistics().getOrThrow()
}

